cmd/relay/bgs/bgs.go

```diff
@@ -0,0 +1,1375 @@
+	// TODO: actually do something with #sync event
+
+	// Broadcast the identity event to all consumers
+	err := bgs.events.AddEvent(ctx, &events.XRPCStreamEvent{
+		RepoSync: evt,
+	})
+	if err != nil {
+		bgs.log.Error("failed to broadcast sync event", "error", err, "did", evt.Did)
+		return fmt.Errorf("failed to broadcast sync event: %w", err)
+	}
+
+	return nil
+}
+
+func (bgs *BGS) handleRepoTombstone(ctx context.Context, pds *models.PDS, evt *comatproto.SyncSubscribeRepos_Tombstone) error {
+	u, err := bgs.lookupUserByDid(ctx, evt.Did)
+	if err != nil {
+		return err
+	}
+
+	if u.PDS != pds.ID {
+		return fmt.Errorf("unauthoritative tombstone event from %s for %s", pds.Host, evt.Did)
+	}
+
+	if err := bgs.db.Model(&User{}).Where("id = ?", u.ID).UpdateColumns(map[string]any{
+		"tombstoned": true,
+		"handle":     nil,
+	}).Error; err != nil {
+		return err
+	}
+	u.SetTombstoned(true)
+
+	return bgs.events.AddEvent(ctx, &events.XRPCStreamEvent{
+		RepoTombstone: evt,
+	})
+}
+
+func (bgs *BGS) purgeDidCache(ctx context.Context, did string) {
+	ati, err := syntax.ParseAtIdentifier(did)
+	if err != nil {
+		return
+	}
+	_ = bgs.didd.Purge(ctx, *ati)
+}
+
+// createExternalUser is a mess and poorly defined
+// did is the user
+// host is the PDS we received this from, not necessarily the canonical PDS in the DID document
+// TODO: rename? This also updates users, and 'external' is an old phrasing
+func (bgs *BGS) createExternalUser(ctx context.Context, did string, host *models.PDS) (*User, error) {
+	ctx, span := tracer.Start(ctx, "createExternalUser")
+	defer span.End()
+
+	externalUserCreationAttempts.Inc()
+
+	bgs.log.Debug("create external user", "did", did)
+	pdid, err := syntax.ParseDID(did)
+	if err != nil {
+		return nil, fmt.Errorf("bad did %#v, %w", did, err)
+	}
+	ident, err := bgs.didd.LookupDID(ctx, pdid)
+	if err != nil {
+		return nil, fmt.Errorf("no ident for did %s, %w", did, err)
+	}
+	if len(ident.Services) == 0 {
+		return nil, fmt.Errorf("no services for did %s", did)
+	}
+	pdsService, ok := ident.Services["atproto_pds"]
+	if !ok {
+		return nil, fmt.Errorf("no atproto_pds service for did %s", did)
+	}
+	durl, err := url.Parse(pdsService.URL)
+	if err != nil {
+		return nil, fmt.Errorf("pds bad url %#v, %w", pdsService.URL, err)
+	}
+
+	if strings.HasPrefix(durl.Host, "localhost:") {
+		durl.Scheme = "http"
+	}
+
+	// TODO: the PDS's DID should also be in the service, we could use that to look up?
+	var peering models.PDS
+	if err := bgs.db.Find(&peering, "host = ?", durl.Host).Error; err != nil {
+		bgs.log.Error("failed to find pds", "host", durl.Host)
+		return nil, err
+	}
+
+	ban, err := bgs.domainIsBanned(ctx, durl.Host)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check pds ban status: %w", err)
+	}
+
+	if ban {
+		return nil, fmt.Errorf("cannot create user on pds with banned domain")
+	}
+
+	if peering.ID == 0 {
```

Comment:
I assume this is old code, maybe from the days when BGS had indexing and would "spider" to discover accounts? or if there is a bulk catch-up happening?

would be good if you could clarify what the expected behavior is here ("maybe this never happens").

Reply:
I did a bunch of reorg and commenting on this function as a whole and I think it's better now